cmd/metal-api/internal/service/machine-service.go

```diff
@@ -944,7 +945,21 @@ func allocateMachine(logger *zap.SugaredLogger, ds *datastore.RethinkStore, ipam
 		}
 	}
 
-	machineCandidate, err := findMachineCandidate(ds, allocationSpec)
+	var machineCandidate *metal.Machine
+
+	err = retry.Do(
+		func() error {
+			var err2 error
+			machineCandidate, err2 = findMachineCandidate(ds, allocationSpec)
```

Comment:
Unfortunately, there is one more complication with the new approach.

After `findMachineCandidate` the `preAllocation` is set to `true`. It can then happen that the `allocateMachine` fails right after finding the machine candidate, for example in in `gatherNetworks`. The pre-allocation is then never set back to false, which makes the machine unusable forever. The rollback mechanism needs to be triggered even after `findMachineCandidate` now.